core/formatter.cpp

```diff
@@ -963,6 +963,280 @@ class PrettyFieldNames : public FmtPass {
     }
 };
 
+/// Add newlines inside complex structures (arrays, objects etc.).
+///
+/// The main principle is that a structure can either be:
+/// * expanded and contain newlines in all the designated places
+/// * unexpanded and not contain any newlines in any of designated places.
```

Comment:
Slightly more concise description:

/// * unexpanded and contain newlines in none of the designated places.